lib/rust/ensogl/core/src/display/scene/layer.rs

```diff
@@ -893,6 +914,50 @@ impl std::borrow::Borrow<LayerModel> for Layer {
 }
 
 
+// === Virtual layers ===
+
+/// A virtual layer within a parent [`Layer`].
+#[derive(Debug, Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct VirtualLayer<S> {
+    layer: WeakLayer,
+    id:    VirtualLayerId,
+    shape: PhantomData<*const S>,
```

Comment:
What does `*const S` mean? (This question is because of my lack of knowledge of this Rust construct just to learn what it does).

EDIT: oh, this is a raw pointer, isn't it? Why are we using raw pointers?

Reply:
Oh, wow, I did not know about it. It's so amazing that you can define the type variable in covariant/contravariant positions this way. This Is such a magic though, I wish Rust authors created a nicer way to define it. Thanks so much for the explanation.